openlibrary/plugins/upstream/borrow.py

```diff
@@ -141,16 +117,26 @@ def POST(self, key):
         if not edition:
             raise web.notfound()
 
-        error_redirect = edition.url("/borrow")
+        # Make a call to availability v2
+        # update the subjects according to result
+        # if `open`, redirect to bookreader
+        olid = key.split('/')[-1]
+        response = lending.get_edition_availability(olid)
+        availability = response[olid] if response else {}
+        ocaid = availability.get('identifier')
+        if availability and availability['status'] == 'open':
+            raise web.seeother('https://archive.org/stream/' + ocaid)
+
+        error_redirect = ('https://archive.org/stream/' + ocaid) if ocaid else edition.url()
         user = accounts.get_current_user()
 
         if user:
             account = OpenLibraryAccount.get_by_email(user.email)
             ia_itemname = account.itemname if account else None
-
         if not user or not ia_itemname:
             web.setcookie(config.login_cookie_name, "", expires=-1)
-            raise web.seeother("/account/login?redirect=%s/borrow" % edition.url())
+            raise web.seeother("/account/login?redirect=%s/borrow?action=%s" % (
+                edition.url(), i.action))
```

Comment:
The only change to this block is the addition of an action (which is from a fixed set)